forecasttools/daily_to_epiweekly.py

```diff
@@ -0,0 +1,68 @@
+from datetime import datetime
+
+import polars as pl
+
+
+# NOTE: this needs to change to ACTUAL epiweeks
+def calculate_epidate(date):
+    iso_calendar = date.isocalendar()
+    return iso_calendar[1], iso_calendar[0]
+
+
+def daily_to_epiweekly(
+    forecast_df: pl.DataFrame,
+    value_col: str = "value",
+    date_col: str = "date",
+    id_cols: list[str] = [".draw"],
+    weekly_value_name: str = "weekly_value",
+    strict: bool = False,
+) -> pl.DataFrame:
+    """
+    Aggregate daily forecast draws to epiweekly.
+    """
+    # check intended df columns are in received df
+    # forecast_df_cols = forecast_df.columns
+    # = [value_col, date_col] + id_cols
+    # assert set(required_cols).issubset(set(forecast_df_cols)),f"Column mismatch between require columns {required_cols} and forecast dateframe columns {forecast_df_cols}."
+    # add epiweek and epiyear columns
+    forecast_df = forecast_df.with_columns(
+        [
+            pl.col(date_col)
+            .map_elements(
+                lambda x: calculate_epidate(datetime.strptime(x, "%Y-%m-%d"))[
+                    0
+                ]
+            )
+            .alias("epiweek"),
+            pl.col(date_col)
+            .map_elements(
+                lambda x: calculate_epidate(datetime.strptime(x, "%Y-%m-%d"))[
+                    1
+                ]
+            )
+            .alias("epiyear"),
+        ]
+    )
```

Comment:
seems like there should be a way to do this without calling `calculate_epidate` twice

Reply:
In R you could use `calculate_epidate` once and split the result using [`unnest_wider`](https://tidyr.tidyverse.org/reference/unnest_wider.html).

``` r
library(tidyverse)


tibble(my_dates = seq(today() - 7, today(), by = "1 day")) |> 
  mutate(epi_dates = map(my_dates,
                         \(x) c(epiyear = epiyear(x), epiweek = epiweek(x)))) |> 
  unnest_wider(epi_dates)
#> # A tibble: 8 × 3
#>   my_dates   epiyear epiweek
#>   <date>       <dbl>   <dbl>
#> 1 2024-09-30    2024      40
#> 2 2024-10-01    2024      40
#> 3 2024-10-02    2024      40
#> 4 2024-10-03    2024      40
#> 5 2024-10-04    2024      40
#> 6 2024-10-05    2024      40
#> 7 2024-10-06    2024      41
#> 8 2024-10-07    2024      41
```

<sup>Created on 2024-10-07 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
For future reference: https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.unnest.html
 
But also I think this will be obviated by replacing `calculate_epidate` with functions from [`epiweeks`](https://epiweeks.readthedocs.io/)

Reply:
It is cool that `datetime` has a built in `isocalendar()` method though. Didn't know that.